docs/question/server.md

```diff
@@ -1,17 +1,88 @@
+
+```python title="server.py"
+import random
+def generate(data):
+    # Generate random parameters for the question and store them in the data["params"] dict:
+    data["params"]["x"] = random.randint(5, 10)
+
+    # Also compute the correct answer (if there is one) and store in the data["correct_answers"] dict:
+    data["correct_answers"]["y"] = 2 * data["params"]["x"]
+```
+
+!!! info
+
+    In general, each function of the question generation process runs *after* all elements in the question. For example, the `generate()` function in `server.py` runs after all elements have run their `generate()` functions. This is important to remember when using the `data` dictionary, as it will contain the results *after* the elements have finished processing.
+
+### Step 2: `prepare`
+
+Next, the `prepare` function is called after all elements (e.g. `<pl-number-input>`) have run `generate()`. This is typically done to do any sort of final post-processing, but is not commonly used.
+
+### Step 3: `render`
+
+Next, the `render(data)` function is called to render the question. You can use this function to override how the question is rendered. This is typically only used for more advanced questions.
+
+### Step 4: `parse`
+
+When a student submits their answer, the `parse` function is called to parse the submitted answers after the individual elements have parsed them. This function can be used to display more-specific format errors than the input elements or to parse the input differently. `<pl-number-input>` will have already parsed the submitted value as an integer, and display an error to the student if it was invalid. For this question, we will only allow the student to submit positive integers, so we will check for that and set a format error with `data["format_errors"]` if it is negative.
+
+```python title="server.py"
+def parse(data):
+    # check we don't already have a format error, and if not, check if the answer is negative
+    if "y" not in data["format_errors"] and data["submitted_answers"]["y"] < 0:
+        data["format_errors"]["y"] = "Negative numbers are not allowed"
+```
+
+### Step 5: `grade`
+
+Finally, the `grade(data)` function is called to grade the question. The grade function is responsible for:
+
+- Setting the score and feedback for each named answer in `data["partial_scores"]`.
+- Setting the total score for the question in `data["score"]`.
+- Setting the overall feedback for the question in `data["feedback"]`.
+
+If this function is not defined, the question will be graded automatically based on the correct answers set in `data["correct_answers"]`. Each answer the student provides will also be given feedback.
+
+It is recommended that you give additional feedback to the student beyond right and wrong, as well as partial credit. If the function is defined, the data you receive has already been graded by the elements. You should make sure that you don't override a correct answer by only giving partial credit.
```

Comment:
I'm not 100% sure what the last sentence here is supposed to mean?

Reply:
If this function is not defined, the question will be graded automatically based on the correct answers set in `data["correct_answers"]`. Each answer the student provides will also be given feedback. If the function _is_ defined, the data you receive has already been graded by the elements. You should ensure you don't lower the score that was already given by the elements. In the example below, we accomplish this using the `marked_as_incorrect` variable, which is set to `True` only if all elements (in this case, there is only one) have marked the named answers as incorrect. In this case, it is safe to award partial credit.

Reply:
Continuing conversation here: https://github.com/PrairieLearn/PrairieLearn/pull/11830/files#r2054784983